### ISSUE TITLE
Dart 2.0: remarks about type inference need adjustment

### ISSUE THREAD

**Issue by chalin** [dart-lang/site-www#360]
In the [dart-2.0 page, lines 107-108](https://github.com/dart-lang/site-www/blame/master/src/dart-2.0.md#L107-L108) it is written that:

> In Dart 2.0, both the analyzer and the runtime try to infer something more specific than `dynamic`.

This sentence needs to be revised because I doubt that there is any runtime type inference. (Type inference is used to attribute _static_ types.)

cc @wmleler @sethladd @kwalrath @kevmoo 

Also, in [sound-dart](https://www.dartlang.org/guides/language/sound-dart#type-inference) it is written:

> When the analyzer can’t infer the type, the `dynamic` type is assigned.

Technically, the analyzer _always_ infers a type, it's just that sometimes it can't infer anything more specific than `dynamic`; i.e., the sentence from the dart-2.0 page quoted above is more accurate, so maybe consider using it (once reference to "runtime" is removed :).

**Comment by chalin** [comment]
Oh, I also noticed something else in the [dart-2.0 page, L110-L112](https://github.com/dart-lang/site-www/blame/master/src/dart-2.0.md#L110-L112):

> While Dart 2.0 is statically typed, type inference and types like `dynamic` and `num` (which can be `int` or `double`) take away much of the burden of specifying types.

But in Dart 1 [`num` could already be an `int` or a `double`](https://api.dartlang.org/stable/1.24.2/dart-core/num-class.html), though the sentence seems to imply otherwise.  In any statically typed language with type inference, it is the type inference itself which alleviates the burden of having to explicitly specify types.

**Comment by Sfshaza** [comment]
I'm not working on Dart for the rest of Q3.

**Comment by kwalrath** [comment]
I'll be taking on Shams's Dart language duties, while she works on Flutter docs for the rest of Q3.

**Comment by chalin** [comment]
I've been giving this some more thought and I think a change in focus maybe be necessary.

Consider [dart-2.0 page, L81](https://github.com/dart-lang/site-www/blame/master/src/dart-2.0.md#L81): _Dart 1.x is an optionally typed language, but types aren't optional in Dart 2.0_.

The focus on "optionality" is at best misleading, and IMHO simply wrong. Let me explain. 

> Dart 1.x is an **optionally typed** language

By "optionally typed" we mean that static type annotations are optional.

> types aren't optional in Dart 2.0

Yes, they still are. I can still write: `var i = 1;`

What _has_ changed is the **type inference algorithm**. It used to be _very_ simple: in the absence of a static type annotation, infer the variable's type to be `dynamic`. One could argue that in Dart 2.0, type inference will **better align with developer expectations**: the type of the initializing expression is actually now used as input to type inference.

In the [Language Tour section on Optional Types](https://www.dartlang.org/guides/language/language-tour#optional-types), it is written:

> **Dart 2.0 note**: Types won’t be optional in Dart 2.0, but you’ll still be able to omit some type annotations, thanks to type inference.

But this is contradictory. I.e.,

- "types won't be optional", said another way: _you **must** write static type annotations_

contradicts

- "but you’ll still be able to omit some type annotations"

(In fact, it is somewhat the opposite: the new type inference in 2.0 will allow developers to safely omit _more_ type annotations.)

**TL;DR** Dart has been and will remain an _optionally typed_ language. But, in Dart 2.0, the new _type inference_ algorithm will allow developers to safely omit type annotations _more often_, and it will allow the analyzer to detect _more type errors_ statically.

So, in Dart 2.0 we're **not** _losing_ anything (optional types), we're _gaining_ something (much better type inference)! Optional types and type inference FTW! :)

Thoughts?

**Comment by lrhn** [comment]
Dart 1 had "optional types" in the sense used by Gilad Bracha in, e.g., [this talk](http://www.bracha.org/pluggable-types.pdf). Type annotations had *no* effect on runtime semantics, they were entirely optional.
There was no inference. The type of `var x;` was not "inferred' to be dynamic - not any more than `int x;` was inferred to be int - it was just what that syntax meant. 

Dart 2 does not have optional types in that sense. Adding types does have a runtime effect. Instead Dart 2 has type inference which allows you to omit types, but that just means the a type is filled in for you.
So, types are required. You can get away with not *writing* the types, but not with not *having* them.

So, in the normal meaning of the term, Dart 2 does not have optional types, it just has type inference.

**Comment by chalin** [comment]
@lrhn, thanks for the feedback.

> Dart 1 had "optional types" in the sense used by Gilad ... Type annotations had no **effect on runtime semantics**, they were entirely optional.

Ah yes! I'd forgotten about that.

> Dart 2 does not have optional types in that sense.

Right.

I'm hoping that we can find an "elevator pitch" summary of the difference between Dart 1 and 2 that doesn't require folks to be intimately familiar with the (unconventional/non-mainstream) semantics of Dart 1.

To achieve this, IMHO, it will help to stick to common meanings: e.g., given that Dart 2 has [type inference](https://en.wikipedia.org/wiki/Type_inference) then, according to Wikipedia (emphasis mine):

> The ability to infer types automatically makes many programming tasks easier, leaving the programmer **free to omit type annotations** while still permitting type checking.

Said another way, "free to omit type annotations" means that type annotations aren't mandatory --
 and so they are optional :).

> So, types are required. ...

From who's point of view? From the p.o.v. of the (internals of the) tooling (e.g. analyzer), then yes.
From the p.o.v. of a developer writing Dart 2.0 code, she/he can still (sometimes) omit type annotations.

**Comment by lrhn** [comment]
I am pedant enough that I will argue against things that are ambiguous and objectively wrong in one of the interpretations :)
* In Dart 2, *type annotations* are optional. 
* In Dart 1, *types* were optional.

So "Dart 2 has optional type annotations" is correct, and so is "Dart 2 to infers types where they are omitted, instead of defaulting to `dynamic`. This improves program analyses without requiring more typing".
Also "Dart 2 has a sound(er) type system which uses the inferred types, and that makes it possible to compile programs more efficiently." (I'm deliberately vague on what that means, in practice it makes DDC work because it trusts the types, and it makes ahead-of-time compilation easier, but it probably doesn't affect JIT compilation at all).

**Comment by sethladd** [comment]
FWIW, the following was a helpful phrasing:

> In Dart 2, type annotations are optional.
> In Dart 1, types were optional.

**Comment by chalin** [comment]
I agree.  The suggested phrasings in https://github.com/dart-lang/site-www/issues/360#issuecomment-325889643 are better.

Building upon @lrhn's words, we could say:

> In Dart 1:
> - **_Types_** were **optional** &mdash; you could erase all type annotations from a program w/o affecting its behavior.

> In Dart 2:
> - **_Types_** are **mandatory** &mdash; they are an essential part of detecting more program errors at compile time, and generating more efficient executables.
> - **_Type annotations_** remain **optional**. In the absence of a type annotation, **_type inference_** is used.

**Comment by sethladd** [comment]
When listing benefits of types (as you do when you're explains "generating more efficient..."), please also mention "improved tooling and feedback when coding" as another benefit of types. (feel free to massage the wording, but developer feedback and improved developer tooling are wins when types are present.

**Comment by kwalrath** [comment]
Thanks for the discussion, @sethladd, @chalin, & @lrhn. I should have a PR for you soon.

**Comment by wmleler** [comment]
I still have a slight problem with this language (being a fellow pedant).
The statements "types are optional" or "types are required" is ambiguous,
because the word "types" has two meanings:

1) types can refer to the static type of an object
2) types can alternatively refer to the type annotations that the
programmer writes

In Dart 2 all objects must have a static type (including "dynamic", which
is a static type that can be dynamic at runtime), which is either supplied
by the programmer or inferred.

**Comment by kwalrath** [comment]
@wmleler Maybe you could do a pull request or send me alternative words?

But let's wait until after my next changes to this page, which will have the latest pub changes to ease the path to 2.0.

**Comment by wmleler** [comment]
I just want to make sure we distinguish between when we are talking about
type annotations, and when we are talking about the types of objects. We
seem to be using the same term ("type") for both, which is what caused this
confusion in the first place. I'm still on the road, but let's talk when I
get back. Maybe this is something for the glossary.

**Comment by chalin** [comment]
Yeah, it is a challenge to find a compact summary of the differences between 1.0 and 2.0 in words that will be the most meaningful to everyday Dart programmers.

> The statements "types are optional" or "types are required" is ambiguous,

Yes, but if we take a bit more of the context from which you quoted those fragments, I think that the meaning of "type" becomes clearer:

> In Dart 1.x, types are optional. You can remove all type annotations from a Dart 1.x program without affecting its behavior.

"Types are optional" was a catch phrase for Dart 1.x, and I believe that @lrhn wants to keep the phrase as is (without extra qualifying words). In any case, the sentence that follows explains the intended meaning: i.e., "types are optional" is meant in the sense that "you can remove all type annotations ... without affecting its behavior". (In programming-language-semantics terminology: Dart's dynamic semantics were completely independent of its static semantics.)

> In Dart 2.0, types are mandatory, but type annotations are still optional. When a type annotation is absent, tools infer the type.

Because of the phrase structure of "types are mandatory, but type annotations are still optional", I feel that it is clear that the first occurrence of "type" isn't referring to a "type annotation". Maybe the second sentence could be a bit more explicit: "When a type annotation is absent from a declaration, tools infer the declaration's **static** type." (The "static" qualification isn't strictly necessary because only static types can be inferred.)

IMHO, the given descriptions are suitable for the intended general audience. I'm not convinced that getting more technical will make things clearer (e.g., stating Dart 2.0's semantics are now conventional, in that we can't even consider the dynamic semantics of a program unless it has first passed the wellformedness checks of the static semantics).

But if you have an alternative (and compact) formulation, that would be great!
